src/compile/data/FilterInvalid.ts

```diff
@@ -0,0 +1,112 @@
+import {SCALE_CHANNELS} from '../../channel';
+import {ScaleType} from '../../scale';
+import {contains, Dict, differ, differArray, duplicate, extend, hash, keys, stringValue} from '../../util';
+import {VgFilterTransform, VgTransform} from '../../vega.schema';
+import {UnitModel} from './../unit';
+import {DataFlowNode} from './dataflow';
+import {isScaleChannel} from '../../channel'
+import {hasContinuousDomain} from '../../scale';
+import {QUANTITATIVE, TEMPORAL} from '../../type';;
+import {FieldDef} from '../../fielddef';
+import {Model, ModelWithField} from '../model';
+
+export class FilterInvalidNode extends DataFlowNode {
+  private _filter_null: Dict<FieldDef<String>>;
+  private _filter_non_pos: Dict<FieldDef<String>>;
+
+  public clone() {
+    return new FilterInvalidNode(extend({}, this._filter_null), extend({}, this._filter_non_pos));
+  }
+
+  constructor(filter_null: Dict<FieldDef<String>>, filter_nonpos: Dict<FieldDef<String>>) {
+    super();
+
+   this._filter_non_pos = filter_nonpos;
+   this._filter_null = filter_null;
+  }
+
+  public static make(model: Model) {
+
+    // create the filter of all the non positive fields
+      const filter_nonpos = SCALE_CHANNELS.reduce(function(nonPositiveComponent, channel) {
+
+        if (model instanceof UnitModel) {
+            const scale = model.getScaleComponent(channel);
+            if (!scale || !model.field(channel)) {
+              // don't set anything
+              return nonPositiveComponent;
+            }
+
+            if (scale.get('type') === ScaleType.LOG) {
+              nonPositiveComponent[model.field(channel)] = model.fieldDef(channel);
+            }
+            return nonPositiveComponent;
+        }
+        return nonPositiveComponent;
+      }, {} as Dict<FieldDef<String>>);
+
+    // create the filter of all the fields that cannot have null
+    let filter_nulls = {};
+    if (model instanceof ModelWithField) {
+      filter_nulls = model.reduceFieldDef((aggregator: Dict<FieldDef<string>>, fieldDef, channel) => {
```

Comment:
As we can only have one dict, you can just reduce all of them together 

`model.reduceFieldDef((aggregator: Dict<FieldDef<string>>, fieldDef, channel)`
and check if `isScaleChannel(channel)` should be pretty optimal.